feat: custom permission types

nextchamp-saqib opened the pull request:
Currently, we can restrict standard actions like read, write, or create on a doctype. With this change, we can also add custom actions such as `download`, `impersonate`, or `execute` and restrict access to relevant functions in code by checking permission for these actions with `frappe.has_permission("User", ptype="impersonate")`. System admins can then allow or block them for roles through the Role Permission Manager.

- [x] make permission type exportable
- [x] apply on docshare
- [x] show custom perm types in permission inspector
- [x] allow creating same permission type for multiple doctypes (eg. 'comment' ptype on ticket & lead)
- [x] add tests
- [x] update docs: https://docs.frappe.io/framework/permission-types

nextchamp-saqib committed:
feat: init `Permission Type` for custom permissions

nextchamp-saqib committed:
feat: extend permission type fields to DocPerm

nextchamp-saqib committed:
feat(temp): `impersonate` permission type for `user` doctype

nextchamp-saqib committed:
Merge branch 'develop' of https://github.com/frappe/frappe into custom-perm-types

nextchamp-saqib committed:
feat: make permission type exportable

nextchamp-saqib committed:
chore: remove unused files

nextchamp-saqib committed:
feat: add label field

nextchamp-saqib committed:
fix: add impersonate perm type

nextchamp-saqib committed:
feat: apply custom permission types on docshare

nextchamp-saqib committed:
fix: allow creation in install

nextchamp-saqib committed:
chore: remove underscore

nextchamp-saqib committed:
refactor: create same perm type for multiple doctypes

nextchamp-saqib committed:
feat: admin can search non-searchable doctypes

nextchamp-saqib committed:
fix: linter issues

nextchamp-saqib committed:
feat: add test cases

nextchamp-saqib committed:
fix: remove duplicates

nextchamp-saqib committed:
fix: linting issues

nextchamp-saqib committed:
fix: cannot create permission types in test

ankush commented on frappe/core/doctype/user/user.json:
```diff
@@ -914,6 +914,7 @@
    "delete": 1,
    "email": 1,
    "export": 1,
+   "impersonate": 1,
```
In migration code, we'll have to migrate `permission type` very early in code and also import perm types before importing doctypes.

nextchamp-saqib replied:
`permission type` is now migrated after `doctype` migration and custom perm types are already imported before any of the app doctypes are synced.

nextchamp-saqib committed:
chore: remove unused import

nextchamp-saqib committed:
fix: handle case if ptype_map for doctype is undefined

nextchamp-saqib committed:
fix: don't change impersonate behavior

- only administrators are allowed to impersonate
- system managers can modify this via role permission manager

nextchamp-saqib committed:
fix: sync permission type doctype before importing custom perm types

nextchamp-saqib committed:
chore: make permission type unsearchable

nextchamp-saqib committed:
feat: show custom perm types in share dialog

nextchamp-saqib committed:
fix: linter errors

nextchamp-saqib committed:
refactor: single perm type for multiple doctypes

problem: if same perm type is defined in two different apps, we'd have to handle the merging of `doc_types` multi-select field

nextchamp-saqib committed:
fix: sync permission types before doctypes

nextchamp-saqib commented:
Known Issue:
For any custom perm type defined on any of the following doctypes, if there's a standard role permission set, they won't be synced at the time of migration.

For eg. if we set role permission on `User` doctype to allow `impersonate` to `System Manager` role, then after migrate, if we check `User` doctype's permission table, we won't see `impersonate` checked

This is because, `User` doctype is synced before any permission types are imported. The same issue persists for any of the doctype synced in this `if` block: https://github.com/frappe/frappe/blob/develop/frappe/model/sync.py#L54
